### DISCUSSION TITLE
[HxGrid] Preserve state with InfiniteScroll navigation mode

### DISCUSSION THREAD

**Discussion by CryptKat** [havit/Havit.Blazor#506] [Q&A] [answered]
Hi! How can I preserve grid state when using InfiniteScroll navigation mode?
There is a page "/orders" with a list of orders. Navigation mode = InfiniteScroll (which is way more convenient than pagination). When the user selects an order, navigation to another page "/orders/{orderId}" happens. When the user returns to the previous page, order list's state (current row index) is lost, and if the user wants to see an adjacent order, he has to scroll down from the beginning.

**Comment by hakenr** [comment, marked as answer]
Hi,
unfortunately, the InfiniteScoll navigation mode is implemented using Blazor's `Virtualize` component and does not support reinstating of the state once it is abandoned. (You might try to use some piece of JavaScript to scroll to the position you save before opening the detail, but I do not recommend going this way.)

What we do on our projects (in such rare cases where we need to preserve InfiniteScroll position when opening and closing the detail) is keeping the `HxGrid` in page (DOM) and visually covering it with the detail UI.

I.e.
* you have to keep the `HxGrid` in DOM
* add the detail to the DOM as the grid-row is selected (you can use HxOffcanvas, HxModal, or just plain `<div>`)
* use appropriate CSS to achieve the visual you need (display the detail "on top of" the grid)
* remove the detail from DOM as you return back to the grid (e.g. with HxCloseButton)

**Reply by CryptKat** [reply]
Thank you. It works like a charm.
I'll leave here my code here, maybe it will be useful for someone.

```razor
@page "/orders"

@if (_currentOrder != null)
{
    <OrderComponent Model="_currentOrder" />
}

<div class="@(_currentOrder != null ? "d-none" : "d-flex (or d-block)")">
    @* Outer Div is necessary for Scoped Css mechanism to work, in my case. 
    Otherwise you can place css classes directly on the HxListLayout *@
    <HxListLayout ...>
        ...
	<DataTemplate>
            <HxGrid @ref="list" SelectedDataItemChanged="HandleSelectedDataItemChanged" />
            ...
            </HxGrid>
        </DataTemplate>
    </HxListLayout>
</div>
```
```csharp
[Parameter, SupplyParameterFromQuery] 
public Guid? Id { get; set; }

[AllowNull] 
private HxGrid<OrderInfo> list;

private Order? _currentOrder;

protected override async Task OnParametersSetAsync()
{
	await base.OnParametersSetAsync();

	if (Id == Guid.Empty)
	{
		_currentOrder = new Order();
	}
	else if (Id != null)
	{
		_currentOrder = await OrderService.GetByIdAsync(Id.Value);
	}
	else
	{
		_currentOrder = null;
		if (list != null) // Necessary to refresh list when closing order (there may be new data).
			await list.RefreshDataAsync();
	}
}

private void HandleSelectedDataItemChanged(OrderInfo selectedDataItem)
{
	Navigation.NavigateTo($"/orders?id={selectedDataItem.Id}");
}
```
